### PR TITLE
chore(docs): Adds graph debug documentation to book

### PR DESCRIPTION
**Pull request by afreeland**
I ran across [this comment](https://github.com/maciejhirsz/logos/issues/265#issuecomment-1939114278) while having some issues with `regex` support.  Thought this was beneficial information and wanted to try and include it in the book so it was a little more visible to people that may be trying to debug and understand how things are working

### CODE REVIEW

File: book/src/debug-graph.md [hunk 1]
```diff
@@ -0,0 +1,116 @@
...
+        . ⇒ 2,
+        [A-Z] ⇒ 4,
+        [a-e] ⇒ 4,
+        f ⇒ 7,
+        [g-z] ⇒ 4,
+    },
+}
+```
+
+Lets look at `@9` for the character `.` we see that it will jump `=>` to `2`.  We can then follow that by looking at `@2` which resolves to our `::Period` token. It will then continue to look for any matches in the case there is potential continuation after the `.` character. In the _input_ we provided there is not, since it is the end of our input.
```
Comment (jeertmans) [inline book/src/debug-graph.md:62]:
```suggestion
This graph can help us understand how our patterns are matched, and maybe understand why we have a bug at some point.

Let's look at the last node, `9`, for the character `.`. We see that it will jump (`=>`) to `2`.  We can then follow that by looking at `2` which resolves to our `::Period` token. It will then continue to look for any matches in the case there is potential continuation after the `.` character. In the _input_ we provided, there is not, since it is the end of our input.
```

Note that I have to read a bit more of the code to feel confident about how Graph actually works.

Also, this may be a bit counter-intuitive to take the example of `.`, where this is the last bit of our input string.

File: book/src/debug-graph.md [hunk 2]
```diff
@@ -0,0 +1,116 @@
...
+
+We also can try to identify how `fast` works by looking at `@9` that `f` jumps to `7`. This will then resolve the last letters of our word _fast_ by matching `ast` which jumps to `8`. Since our provided _input_ to the lexer does not include alphanumeric character after the word "fast" but rather whitespace, the token `::Fast` will be recognized. Then the graph will look for further potential continuation (ie `[g-z] => 4`)
```
Comment (jeertmans) [inline book/src/debug-graph.md:64]:
```suggestion
We also can try to identify how `fast` works by looking at `9`, first, and seeing that `f` makes it jump to `7`. This will then resolve the last letters of our word _fast_ by matching `ast` which jumps to `8`. Since our provided _input_ to the lexer does not include alphanumeric character after the word "fast" but rather whitespace, the token `::Fast` will be recognized. Then the graph will look for further potential continuation (i.e., `[g-z] => 4`)
```

File: book/src/debug-graph.md [hunk 3]
```diff
@@ -0,0 +1,116 @@
...
+Run `cargo build` 
+
+
+### Step 4
+Now in your code that uses Logos, point your Logos dependency in `Cargo.toml` to your local repo
+```
+[dependencies]
+logos =  { path = "../path/to/logos" }
+regex = "1.10.3"
+```
```
Comment (jeertmans) [inline book/src/debug-graph.md:116]: I'd be happy if there was another way than cloning the repository, because this feels overly complicated for the basic user ^^'

If not already, it might be good to add a feature that enables debugging, like with Clap's debug feature: https://github.com/clap-rs/clap/blob/0c01b5558de0a1a513a2af429981099e550d1a78/Cargo.toml#L77.

File: book/src/debug-graph.md (hunk 1 above)
Comment (jeertmans) [reply book/src/debug-graph.md:62]: The thing I was to know (I guess the information is somewhere in the source code) is whether we always enter the graph via its last entry, or not.

**Review by jeertmans — CHANGES_REQUESTED** [review]
Hey! Thank you some much for your contribution! I left a few review comments that I'd like you to address before we can merge :-)

Feel free to accept or reject my suggestions!

File: book/src/debug-graph.md (hunk 1 above)
Comment (afreeland) [reply book/src/debug-graph.md:62]: Totally agree about it being counter-intuitive about the `.` example.  I used the `.` since it seemed a little more straight forward on the graph to understand.  Could probably smooth over that some if I reworked some of the wording to lead into a bit more, instead of BLAM, understand this lol

**Comment by afreeland** [comment]
Yeah, sorry for the kind of rushed PR...trying to sneak this in-between family time lol

Made some improvements to language and file structure based off of your recommendations.  Hopefully, a little less jarring to someone reading it.

As for the enable debugging part, I totally agree, I am new to rust but I can try to put something together that properly uses a `debug-mode` kind of flag.  Will keep it as a separate commit or could be a follow up PR if I get a chance.  Would make life much nicer if something was an easy `feature` bit to flip.

File: book/src/debug-graph.md (hunk 3 above)
Comment (afreeland) [reply book/src/debug-graph.md]: I looked into this some, trying to add a _feature_ that could easily be toggled.  Allowing a user to run Logos with the a `debug` feature enabled it to output the graph.  

I tried adding `debug` to the `Cargo.toml` in the root of the Logos repo
```
// Cargo.toml in repo root
[features]
debug = []
```
Then built with
`cargo build --features debug`

Then attempted to use that in another project like this
```
[dependencies]
logos =  { path = "/Users/afreeland/git/af-logos", features = ["debug"] }
```

However, that doesn't appear work because the configuration actually gets used in the `logos-codegen` section. Which is down a level and with its own `Cargo.toml` file. 

The other complication is `logos-derive` actually has `logos-codegen` as a dependency...so I tried adding the `debug` feature here, to see if it would propagate down the chain

```
// logos-derive/Cargo.toml
[dependencies]
logos-codegen = {version = "0.14.0", path = "../logos-codegen", features = ["debug"]}
```

This did output the graph when running from a different package outside of Logos but it was ALWAYS enabled at this point and not propagating.

Perhaps there is a way to pass from the top down but haven't discovered it yet.

Comment (jeertmans) [reply book/src/debug-graph.md]: I actually just added a `debug` feature, so you can use that instead.

See #382.

Once merged, you'll be able to rebase your branch and use this option (instead of cloning the git and modifying the source code) :-)

**Comment by jeertmans** [comment]
> Yeah, sorry for the kind of rushed PR...trying to sneak this in-between family time lol
> 
> Made some improvements to language and file structure based off of your recommendations. Hopefully, a little less jarring to someone reading it.
> 
> As for the enable debugging part, I totally agree, I am new to rust but I can try to put something together that properly uses a `debug-mode` kind of flag. Will keep it as a separate commit or could be a follow up PR if I get a chance. Would make life much nicer if something was an easy `feature` bit to flip.

No issue, I did it myself since this was an easy fix. The issue you had was that you needed to declare a `debug` feature in sub-crates. See #382 for more details :-)

File: book/src/debugging.md [hunk 4]
```diff
@@ -0,0 +1,127 @@
+# Debugging
+
+Gain deeper insights into your code's behavior with this debugging section. Here you can discover techniques and tools to explore and understand your project's internals.
+
+## Visualizing Logos Graph 
+
+Logos works by creating a graph that gets derived from the tokens that you create.  This graph visualizes how the lexer moves through different states when processing input.
```
Comment (jeertmans) [inline book/src/debugging.md:7]:
```suggestion
Logos works by creating a graph that gets derived from the tokens that you defined. This graph describes how the lexer moves through different states when processing input.
```

Comment (jeertmans) [reply book/src/debugging.md:7]: Nice description!

File: book/src/debugging.md [hunk 5]
```diff
@@ -0,0 +1,127 @@
...
+
+It may be beneficial during debugging to be able to visualize how Logos has generated its graph.
```
Comment (jeertmans) [inline book/src/debugging.md:9]:
```suggestion
Hence, it may be beneficial during debugging to be able to visualize this graph,
to understand how Logos will match the various tokens. 
```

File: book/src/debugging.md [hunk 6]
```diff
@@ -0,0 +1,127 @@
...
+        f ⇒ 7,
+        [g-z] ⇒ 4,
+    },
+}
+```
+This graph can help us understand how our patterns are matched, and maybe understand why we have a bug at some point.
+
+Let's get started by trying to understand how Logos is matching the `.` character, which we've tokenized as `Token::Period`.
+
+We can begin our search by looking at number `@9` for the character `.`. We can see that if Logos matches a `.` it will jump `=>` to number `2`.  We can then follow that by looking at `@2` which resolves to our `::Period` token.
```
Comment (jeertmans) [inline book/src/debugging.md:71]: I am not sure where the `@` number comes from (maybe from the comment you mentioned), but it looks weird to me to add `@` before the number, when there is none in the debug message.

Comment (jeertmans) [reply book/src/debugging.md:71]: Also, you then say number `2`, then `@2`, which may be confusing

File: book/src/debugging.md [hunk 7]
```diff
@@ -0,0 +1,127 @@
...
+    },
+}
+```
+This graph can help us understand how our patterns are matched, and maybe understand why we have a bug at some point.
+
+Let's get started by trying to understand how Logos is matching the `.` character, which we've tokenized as `Token::Period`.
+
+We can begin our search by looking at number `@9` for the character `.`. We can see that if Logos matches a `.` it will jump `=>` to number `2`.  We can then follow that by looking at `@2` which resolves to our `::Period` token. 
+
+Logos will then continue to look for any matches past our `.` character.  This is required in case there is potential continuation after the `.` character. Although, in the _input_ we provided there are not any additional characters, since it is the end of our input.
```
Comment (jeertmans) [inline book/src/debugging.md:73]:
```suggestion
Logos will then continue to look for any matches past our `.` character.  This is required in case there is potential continuation after the `.` character. Although, in the _input_ we provided there are no any additional characters, since it is the end of our input.
```

File: book/src/debugging.md [hunk 8]
```diff
@@ -0,0 +1,127 @@
...
+
+We also can try to identify how the token `fast` works by looking at `@9`, first, and seeing that `f` will cause Logos to jump to `7`. This will then resolve the last letters of our word _fast_ by matching `ast` which jumps to `@8`. Since our provided _input_ to the lexer does not include alphabetic characters after the word "fast" but rather whitespace, the token `::Fast` will be recognized. Then the graph will look for further potential continuation (i.e., `[g-z] => 4`)
```
Comment (jeertmans) [inline book/src/debugging.md:75]:
```suggestion
We also can try to identify how the token `fast` works by looking at `@9`, first, and seeing that `f` will cause Logos to jump to `7`. This will then resolve the last letters of our word _fast_ by matching `ast` which jumps to `@8`. Since our provided _input_ to the lexer does not include alphabetic characters after the word "fast", but rather a whitespace, the token `::Fast` will be recognized. Then, the graph will look for further potential continuation (here, `[g-z] => 4`)
```

File: book/src/debugging.md [hunk 9]
```diff
@@ -0,0 +1,127 @@
...
+
+### Enabling
```
Comment (jeertmans) [inline book/src/debugging.md:77]: This and the following section should be one level above, so:

```markdown
## Enabling

### Step 1...
```

Also, maybe name the steps (instead of saying step 1, step 2). But this may change as a result of #382.

**Review by jeertmans — CHANGES_REQUESTED** [review]
Nice work @afreeland! I left some more comments (sorry to be picky about some ^^'), but I hope you don't mind..

Note that #382 was merged and just added a new debug feature.

File: book/src/debugging.md [hunk 10]
```diff
@@ -0,0 +1,127 @@
+# Debugging
+
+Gain deeper insights into your code's behavior with this debugging section. Here you can discover techniques and tools to explore and understand your project's internals.
```
Comment (marcospb19) [inline book/src/debugging.md:3]: I'd reword this, I think this is about debugging Logos's lexer behavior and not my project internals.

File: book/src/debugging.md (hunk 6 above)
Comment (afreeland) [reply book/src/debugging.md:71]: Yeah, I was just keeping the `@` from the comment but went ahead and stripped the `@` symbol

File: book/src/debugging.md (hunk 9 above)
Comment (afreeland) [reply book/src/debugging.md:77]: I rebased and provided a simplified example that leverages your `debug` feature.

Comment (marcospb19) [reply book/src/debugging.md:77]: I also believe that "# Enabling" should be the first section.

File: book/src/debugging.md [hunk 11]
```diff
@@ -0,0 +1,85 @@
+# Debugging
+
+Gain deeper insights into your code's behavior with this debugging section.
```
Comment (marcospb19) [inline book/src/debugging.md:3]:
```suggestion
Instructions on how to debug your Logos lexer.
```

**Comment by abhillman** [comment]
Users wishing to test this functionality may do so, pending this patch making it upstream to `crates.io`, with the following:

```
logos = { git = "https://github.com/maciejhirsz/logos.git#afreeland:debug-graph" }
```

**Comment by abhillman** [comment]
I haven't had a chance to dive in yet, but is there a public API for fetching, manipulating, and constructing graphs exist besides using macros? Some reasons I can think of why those APIs would be useful:

**fetching**
* if a project contains multiple lexers, being able to fetch a graph and print could be easier for users than printing all graphs, as is done in this PR
* in many cases, such an API would enable a developer to easily use the `logos` macro DSL to produce a lexer file for other tools (flex and its many derivations, for example)

**manipulating/construction**:
* although this may require additional changes, this would allow for producing lexers at runtime

**Comment by jeertmans** [comment]
> I haven't had a chance to dive in yet, but does logos have a runtime-accessible public API for fetching, manipulating, and constructing graphs exist besides using macros? Some reasons I can think of why those APIs would be useful:
> 
> **fetching** [related to this PR]
> 
> * if a project contains multiple lexers, being able to fetch a graph and print could be easier for users than printing all graphs, as is done in this PR
> * in many cases, such an API would enable a developer to easily use the `logos` macro DSL to produce a lexer file for other tools (flex and its many derivations, for example)
> 
> **manipulating/construction** [out of scope for this PR, yet topically related]:
> 
> * although this may require additional changes, this would allow for producing lexers at runtime
> 
> **What this PR might look like if there is/were a fetch API**
> 
> The documentation would mention something like this:
> 
> ```rust
> let graph = Token::graph();
> println!("{:#?}", graph)
> ```
> 
> Continuing to use a feature flag to enable `::graph()` (if something like it doesn't already exist) would still also probably make sense.
> 
> Thoughts?

This is true that is might be nice to expose that as part of the public API. Would you mind creating an issue for that feature request?

**Review by jeertmans — APPROVED** [review]
